### PR TITLE
New features and bug fixes: dispensations & grades

### PR DESCRIPTION
**Pull request by R3dlessX**
# Description

- Ajout des "dispenses"
- Correction d'un bug au niveau de la gestion du type de note (abs, disp..)

# Informations supplémentaires: 
OS: MacOS Sequoia

# Tests
Capture d’écran 2024-12-11 à 02 07 27

```
mac@Air-de-Mac-5 Pawdirecte % tsx ./examples/student-attendance.ts 
Initializing a session using credentials...
Double authentication required.
Reply to this question: Quel est votre mois de naissance ?
[0] janvier
[1] février
[2] mars
[3] avril
[4] mai
[5] juin
[6] juillet
[7] août
[8] septembre
[9] octobre
[10] novembre
[11] décembre
Answer the question by providing the index of the answer: X
Double authentication confirmed.
Logged in to XX XX from XX
{
  punishments: [],
  absences: [
    {
      id: 210,
      studentId: 4331,
      studentName: '',
      reason: 'Retards des transports scolaires',
      date: 2024-09-05T00:00:00.000Z,
      dateOfEvent: Invalid Date,
      label: '00:10',
      teacher: '',
      comment: '',
      justified: true,
      justificationType: '',
      onlineJustification: false,
      todo: '',
      kind: 'Retard',
      displayDate: 'le jeudi 05 septembre 2024 de 08:10 à 08:20'
    }
  ],
  exemptions: [
    {
      id: 92,
      studentId: 4331,
      studentName: 'XX XXX',
      reason: '',
      date: 2024-09-02T00:00:00.000Z,
      dateOfEvent: Invalid Date,
      label: '',
      teacher: '',
      comment: '',
      justified: false,
      justificationType: '',
      onlineJustification: false,
      todo: '',
      kind: 'Dispense',
      displayDate: 'du lundi 02 septembre 2024 à 08:10 au mercredi 30 juillet 2025 à 10:20'
    }
  ]
}

### CODE REVIEW

File: src/models/attendance-item.ts [hunk 1]
```diff
@@ -10,6 +10,7 @@ export type AttendanceItem = Readonly<{
   label: string;
   teacher: string;
   comment: string;
+  subjectName: string;
```
Comment (Vexcited) [inline src/models/attendance-item.ts:13]: I don't see the `subjectName` property in the example outputs you gave :

```js
exemptions: [
  {
    id: 92,
    studentId: 4331,
    studentName: 'XX XXX',
    reason: '',
    date: 2024-09-02T00:00:00.000Z,
    dateOfEvent: Invalid Date,
    label: '',
    teacher: '',
    comment: '',
    justified: false,
    justificationType: '',
    onlineJustification: false,
    todo: '',
    kind: 'Dispense',
    displayDate: 'du lundi 02 septembre 2024 à 08:10 au mercredi 30 juillet 2025 à 10:20'
  }
]
```

Is that needed ?

File: src/decoders/grade-value.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ export const decodeGradeValue = (value: string): GradeValue => {
       points: 0
     };
 
-  switch (value) {
+  switch (value.trim()) {
```
Comment (Vexcited) [inline src/decoders/grade-value.ts:10]: Is this also needed ? It's not trimmed already on EcoleDirecte's side ?

**Review by Vexcited — COMMENTED** [review]
LGTM, there's a few questions I have though before merging.

File: src/models/attendance-item.ts (hunk 1 above)
Comment (R3dlessX) [reply src/models/attendance-item.ts:13]: I've just corrected the problem, it's used to display the name of the subject we're exempt from.

```
{
  punishments: [],
  absences: [
    {
      id: 210,
      studentId: 4331,
      studentName: '',
      reason: 'Retards des transports scolaires',
      date: 2024-09-05T00:00:00.000Z,
      dateOfEvent: Invalid Date,
      label: '00:10',
      teacher: '',
      comment: '',
      justified: true,
      justificationType: '',
      onlineJustification: false,
      subjectName: '',
      todo: '',
      kind: 'Retard',
      displayDate: 'le jeudi 05 septembre 2024 de 08:10 à 08:20'
    }
  ],
  exemptions: [
    {
      id: 92,
      studentId: 4331,
      studentName: 'XXX XXX',
      reason: '',
      date: 2024-09-02T00:00:00.000Z,
      dateOfEvent: Invalid Date,
      label: '',
      teacher: '',
      comment: '',
      justified: false,
      justificationType: '',
      onlineJustification: false,
      subjectName: 'ED.PHYSIQUE & SPORT.',
      todo: '',
      kind: 'Dispense',
      displayDate: 'du lundi 02 septembre 2024 à 08:10 au mercredi 30 juillet 2025 à 10:20'
    }
  ]
}

File: src/decoders/grade-value.ts (hunk 2 above)
Comment (R3dlessX) [reply src/decoders/grade-value.ts:10]: Yes, it's not trimmed in the EcoleDirecte side. Exemple for the grades:
Capture d’écran 2024-12-11 à 10 07 25

The value is “Abs ” and not “Abs"

**Pull request merged**